### PR TITLE
BC-5667 - refactor content element bar

### PR DESCRIPTION
**Pull request by hoeppner-dataport**
# Short Description
Refactoring of UI-Board: Content-Element-Bar

## Links to Ticket and related Pull-Requests
https://ticketsystem.dbildungscloud.de/browse/BC-5667

## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team
- [x] DEV: Every new component is implemented having accessibility in mind (e.g. aria-label, role property)

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

File: src/components/feature-board-link-element/components/LinkContentElement.vue [hunk 1]
```diff
@@ -2,13 +2,15 @@
...
 		data-testid="board-link-element"
+		ref="linkContentElement"
 		dense
 		elevation="0"
 		:outlined="outlined"
-		ref="linkContentElement"
 		:ripple="false"
 		tabindex="0"
 		@keydown.up.down="onKeydownArrow"
+		:href="sanitizedUrl"
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-link-element/components/LinkContentElement.vue:12]: We talked about this. This change was necessary for the linkElement to allow proper keyboard-navigation. It does not have to be the same for the linkElement.

File: src/components/feature-board-drawing-element/InnerContent.vue [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	name: "InnerContent",
+	components: { ContentElementBar },
+	props: {
+		lastUpdatedAt: {
+			type: String,
+			required: true,
+		},
+	},
+	setup(props) {
+		const i18n = injectStrict(I18N_KEY);
```
Comment (uidp) [inline src/components/feature-board-drawing-element/InnerContent.vue:37]: Would be nicer to use the `useI18n()` composable because easier to migrate to Vue3.

File: src/components/feature-board-drawing-element/InnerContent.vue [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+			<v-img :src="imageSrc" height="185px" alt="" cover />
+		</template>
+		<template #title>
+			{{ $t("components.cardElement.drawingElement") }}
+		</template>
+		<template #subtitle>
+			{{ $t("components.cardElement.lastUpdatedAt") }}
+			{{ formattedLastUpdatedAt }}
+		</template>
+		<template #menu v-if="$slots.default">
```
Comment (uidp) [inline src/components/feature-board-drawing-element/InnerContent.vue:13]: Can there really be a situation where the default slot is not defined?

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 4]
```diff
@@ -35,7 +52,11 @@ export default defineComponent({
 		hasGreyBackground: { type: Boolean, required: false },
```
Comment (uidp) [inline src/components/ui-board/content-element/ContentElementBar.vue:52]: I think we should pick a more "semantic" property name here and not include the color "grey".

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 5]
```diff
@@ -1,32 +1,49 @@
...
+				<ContentElementTitleIcon v-if="icon" :icon="icon" class="mr-2" />
+
+				<ContentElementTitle>
+					<slot name="title" />
+				</ContentElementTitle>
 
-			<slot name="element" />
+				<div v-if="$slots.menu && !$slots.display" class="three-dot-menu">
+					<slot v-if="$slots.menu && !$slots.display" name="menu" />
+				</div>
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue]: We need this condition in order to have it not in each component.
By this we are able to 
1. always place the menu on "the display" e.g. the preview image, when it is available.
2. always place the menu next to the headline if it is available but nothing is there for the display-slot.
Otherwise each component would need to implement that.

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 6]
```diff
@@ -35,7 +52,11 @@ export default defineComponent({
...
 	},
-	components: { ContentElementTitleIcon },
+	components: {
+		ContentElementDisplay,
+		ContentElementTitleIcon,
+		ContentElementTitle,
+	},
 	setup(props) {
 		const backgroundClass = computed(() => {
 			return props.hasGreyBackground ? "grey lighten-4" : "";
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:62]: I would suggest that you extend the contentElementBar inside the audio-element ticket. Or the requirements need to fit better to the ui-concept for elements. That's something UXies and POs have to clearify in my opinion.

File: src/components/feature-board-file-element/content/display/file-description/FileDescription.vue [hunk 7]
```diff
@@ -6,10 +6,9 @@
 			:icon="mdiFileDocumentOutline"
 		>
 			<template #title>
-				<ContentElementTitle class="title-link">
-					<a v-if="src" :href="src" target="_blank">{{ name }}</a>
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-file-element/content/display/file-description/FileDescription.vue:10]: That's right: until now the fileElement did handle these differently - but that's an inconsistency that also led to a fileDescription that seems to have an action - but does not have.
I think I discussed this situation with Anna and the UXies decide to be here more consistent.
But to be honest it would now be technical possible to rollback this change, as the tooltip is now a vuetify-tooltip and is now capable of displaying the <a>-tag, too.
tldr: if title content is more simple it is easier to have the required long-title behaviour with ellipsis and tooltip.

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 8]
```diff
@@ -47,3 +68,34 @@ export default defineComponent({
...
+.content-element-bar:hover .content-element-bar-display {
+	filter: brightness(80%);
+}
+.content-element-bar .subtitle,
+.content-element-bar .description {
+	line-height: 20px;
+}
+.content-element-bar .three-dot-menu {
+	margin-top: -6px;
+	margin-right: -6px;
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:89]: Because it would not work here.
Menubutton is much bigger than it's icon, as it has a much bigger round area only visible on hover.
The UX-design and it's positioning and everything does not respect that.

File: src/components/feature-board-file-element/content/footer/download/FileDownload.vue [hunk 9]
```diff
@@ -34,3 +34,9 @@ export default defineComponent({
 	},
 });
 </script>
+
+<style type="text/scss">
+.download-button {
+	margin-right: -6px;
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-file-element/content/footer/download/FileDownload.vue:40]: True. If there ever exists a reason to change the positioning of the 3dot menu - this place needs to be adapted, too.

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 10]
```diff
@@ -1,32 +1,49 @@
...
+			</div>
 
-			<v-spacer />
+			<div v-if="$slots.element" class="content-element-bar-element">
+				<slot name="element" />
+			</div>
 
-			<slot name="menu" />
-		</v-app-bar>
+			<div v-if="$slots.subtitle" class="pt-1 gb-1 subtitle">
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:31]: No - it's not the same thing.
A description is a clear naming in my opinion it will only hold a description - if one is available.
The name subtitle defines more the position - then the content (= will be below the title and maybe detailing it).
This slot will be used for different types of information - depending on the element-type.
I think this is useful and better than having 5 additional slots having a different name but the same style and being always used alone.

File: src/components/ui-board/content-element/ContentElementBar.vue (hunk 5 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue]: My hope is that the generalization would help here, too, to unify the behaviour.
Would be great if new elements or situations priorize the found generalisation in the UX-team over custom solutions for individual elements.
If the outcome is not this way, it is okay to enable this flexibilty by adding a configuration option (e.g. "fixedMenuPostion=true" or something similar/better).

File: src/components/ui-board/content-element/ContentElementBar.vue (hunk 8 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:89]: Different point, but sounds valid - will change that.

File: src/components/feature-board-submission-element/components/SubmissionContentElementEdit.vue [hunk 11]
```diff
@@ -1,11 +1,14 @@
 <template>
 	<div>
-		<div class="grey lighten-4 pa-4 rounded">
-			<SubmissionContentElementTitle />
-			<div class="menu">
+		<ContentElementBar :hasGreyBackground="true" :icon="mdiLightbulbOnOutline">
+			<template #title>
+				{{ t("components.cardElement.submissionElement") }}
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-submission-element/components/SubmissionContentElementEdit.vue:5]: Then I get an error for:
:date-input-label="$t('common.labels.date')"
:time-input-label="$t('common.labels.time')"
due to $t might return undefined--- theoretically.

[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-submission-element/components/SubmissionContentElementEdit.vue:5]: done

File: src/components/ui-board/index.ts [hunk 12]
```diff
@@ -16,6 +17,7 @@ export {
 	BoardMenuActionMoveUp,
 	BoardMenuActionMoveDown,
 	ContentElementBar,
+	ContentElementDisplay,
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/index.ts:20]: Right - would be great if we only export ContentElementBar und stop exporting any of the subcomponents (e.g. ComponentElementTitle and ComponentElementTitleIcon).
fixed.

File: src/components/feature-board-file-element/content/display/file-description/FileDescription.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-file-element/content/display/file-description/FileDescription.vue:10]: Have alook at BC-5996

File: src/components/feature-board-file-element/content/footer/download/FileDownload.vue (hunk 9 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-file-element/content/footer/download/FileDownload.vue:40]: Agreed with Max that Dune will revisit this with other ticket. (together with the thread above)

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 13]
```diff
@@ -1,32 +1,49 @@
 <template>
-	<div :class="backgroundClass">
-		<v-app-bar dense flat color="transparent">
-			<ContentElementTitleIcon v-if="icon" :icon="icon" class="mr-2" />
+	<div tabindex="-1" class="content-element-bar">
+		<div v-if="$slots.display">
+			<ContentElementDisplay>
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:4]: Idea: Renaming to ContentElementLayout.
Agreed with Max to not rename it now and to rediscuss it between Devs.

File: src/components/ui-board/content-element/ContentElementTitle.vue [hunk 14]
```diff
@@ -1,11 +1,90 @@
...
 </script>
+
+<style lang="scss" scoped>
+.content-element-title.subtitle-1 {
+	-webkit-text-overflow: ellipsis;
+	-o-text-overflow: ellipsis;
+	text-overflow: ellipsis;
+	text-rendering: optimizelegibility;
+	overflow: hidden;
+	font-family: var(--font-accent) !important;
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementTitle.vue:62]: Agreed with max, that it would be great to have a <h4> here that uses the centrally defined defaults.
Generally it would great to have all headings cleanly defined so that we can use them for Page aka Board-Title, columnTitle and CardTitle, too. Therefore resolved for this ticket.

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 15]
```diff
@@ -1,32 +1,49 @@
 <template>
-	<div :class="backgroundClass">
-		<v-app-bar dense flat color="transparent">
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:3]: Might work, but maybe not due to our margin-requirements for the 3dotmenu.
Agreed to not work on it in this ticket.

File: src/components/ui-board/content-element/ContentElementTitle.vue [hunk 16]
```diff
@@ -1,11 +1,90 @@
...
+	>
+		<v-tooltip
+			top
+			start
+			origin="auto"
+			open-delay="300"
+			close-delay="300"
+			transition="fade"
+			content-class="tooltip"
+			v-if="isOverflowingLongText"
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementTitle.vue:15]: Agreed with Max that we need to define that in FrontendExchange-Group together with other test-specific questions. And that it is a teamwise deciseion until that.

File: src/components/feature-render-html/RenderHTML.vue [hunk 17]
```diff
@@ -1,9 +1,13 @@
 <template>
-	<component :is="component" v-dompurify-html:[config]="html" />
+	<component
+		:is="component"
+		v-if="html !== ''"
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-render-html/RenderHTML.vue:4]: It's probably just a refactoring of renderHTML but uses the same code. Will have a look anyway.

File: src/components/feature-board-link-element/components/LinkContentElementDisplay.vue [hunk 18]
```diff
@@ -1,44 +1,36 @@
...
-				</template>
-				<template #subtitle>
-					{{ hostname }}
-				</template>
-			</ContentElementBar>
-		</a>
-	</v-card>
+		<ContentElementBar :hasGreyBackground="true" :icon="mdiLink">
+			<template #display>
+				<v-img v-if="imageUrl" :src="imageUrl" alt="" class="rounded-t" />
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-link-element/components/LinkContentElementDisplay.vue:11]: Good idea. I think not all refactorings of all elements could be done in this ticket. So: it should be done, too - to benefit from this new previewImage component - but not here ;-) Agreed with max.

File: src/components/feature-board-drawing-element/InnerContent.vue [hunk 19]
```diff
@@ -0,0 +1,52 @@
+<template>
+	<ContentElementBar :hasGreyBackground="true" :icon="mdiPresentation">
+		<template #display>
+			<v-img :src="imageSrc" height="185px" alt="" cover />
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-drawing-element/InnerContent.vue:4]: Same as above.

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 20]
```diff
@@ -1,32 +1,49 @@
...
-				inactive
-			>
-				<slot name="title" />
-			</v-toolbar-title>
+				<template #menu>
+					<slot name="menu" />
+				</template>
+			</ContentElementDisplay>
+		</div>
+		<div :class="backgroundClass" class="pa-4 rounded-b">
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:14]: Problem: more space than before, if suggested conditions are not met. (have a look at FileElement type image without description).

File: src/components/feature-board-file-element/content/inputs/FileInputs.vue [hunk 21]
```diff
@@ -1,5 +1,5 @@
 <template>
-	<v-card-text v-if="isEditMode">
+	<div v-if="isEditMode" class="px-4">
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-file-element/content/inputs/FileInputs.vue:2]: Extended the padding in FileInput for the moment. Should be correct automatically when using "display"-slot here, too. Which will be part of the coming refacotrings in the file-element.
image

File: src/components/feature-board-file-element/content/display/file-description/FileDescription.vue [hunk 22]
```diff
@@ -45,7 +44,7 @@ export default defineComponent({
 			mdiFileDocumentOutline,
 		};
 	},
-	components: { ContentElementBar, ContentElementTitle },
+	components: { ContentElementBar },
 });
 </script>
 <style scope>
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-file-element/content/display/file-description/FileDescription.vue:50]: Removed unneeded text-link styles. (And the div that assign the text-link-class)

File: src/components/feature-board-drawing-element/InnerContent.vue (hunk 2 above)
Comment (hoeppner-dataport) [reply src/components/feature-board-drawing-element/InnerContent.vue]: True. fixed. :)

File: src/components/feature-board-drawing-element/InnerContent.vue (hunk 3 above)
Comment (hoeppner-dataport) [reply src/components/feature-board-drawing-element/InnerContent.vue:13]: Tested: without the if it still behaves the same way.
Removed.
Fixed.

File: src/components/ui-board/content-element/ContentElementBar.vue (hunk 4 above)
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:52]: That's a valid point but will not be change in this PR. 
Dune is planing to extend the flexibility of the component in this regard.

File: src/components/feature-board-link-element/components/LinkContentElementCreate.vue [hunk 23]
```diff
@@ -1,5 +1,5 @@
 <template>
-	<v-card-text>
+	<v-card class="pa-4 elevation-0" tabindex="-1">
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-link-element/components/LinkContentElementCreate.vue:2]: v-card-text would add some inner-spacing that would be incorrect here.

File: src/components/feature-board-submission-element/components/SubmissionContentElementEdit.vue [hunk 24]
```diff
@@ -1,11 +1,14 @@
 <template>
 	<div>
-		<div class="grey lighten-4 pa-4 rounded">
-			<SubmissionContentElementTitle />
-			<div class="menu">
+		<ContentElementBar :hasGreyBackground="true" :icon="mdiLightbulbOnOutline">
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-submission-element/components/SubmissionContentElementEdit.vue:3]: Looked into it:
Submission element gets now an underlined title on hover - although no action can be executed here:
image
(left: main | right: this branch)
UX input needed - or should be merged this way?

File: src/components/feature-board-link-element/components/LinkContentElementCreate.vue (hunk 23 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-link-element/components/LinkContentElementCreate.vue:2]: rolled back.
fixed.

File: src/components/ui-board/content-element/ContentElementTitleIcon.vue [hunk 25]
```diff
@@ -1,8 +1,12 @@
 <template>
 	<div>
-		<v-icon color="black" size="18" data-testid="content-element-title-icon">{{
-			icon
-		}}</v-icon>
+		<v-icon
+			class="content-element-title-icon"
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementTitleIcon.vue:4]: Is used now to align the icon.
solved.

File: src/components/feature-render-html/RenderHTML.vue (hunk 17 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-render-html/RenderHTML.vue:4]: rolled back these changes, as they should not be part of this ticket

File: src/components/ui-board/content-element/ContentElementBar.vue (hunk 20 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue:14]: hide content-block if only display-slot has content
fixed

File: src/components/feature-board-file-element/content/display/file-description/FileDescription.vue (hunk 7 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-file-element/content/display/file-description/FileDescription.vue:10]: BC-5996 should be discussed on Tuesday, too - in order to ensure that it is completely valid.
For the moment I added the a-link back in order to ensure, that clicking on the title of a fileelment behaves the same as before the PR.

File: src/components/ui-board/content-element/ContentElementBar.vue (hunk 20 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue]: Added the suggested conditions.
fixed.

File: src/components/feature-board-submission-element/components/SubmissionContentElementEdit.vue (hunk 24 above)
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/feature-board-submission-element/components/SubmissionContentElementEdit.vue]: Reverted - due to unclearity and open fundamental changes for the submission element.

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 26]
```diff
@@ -1,32 +1,49 @@
...
 
-			<slot name="element" />
+				<div v-if="$slots.menu && !$slots.display" class="three-dot-menu">
+					<slot name="menu" />
+				</div>
+			</div>
 
-			<v-spacer />
+			<div v-if="$slots.element" class="content-element-bar-element">
+				<slot name="element" />
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue]: fixed

File: src/components/ui-board/content-element/ContentElementBar.vue [hunk 27]
```diff
@@ -47,3 +68,34 @@ export default defineComponent({
...
+	margin-top: -6px;
+	margin-right: -6px;
+}
+.content-element-bar .three-dot-menu .v-icon {
+	color: var(--v-black-base);
+}
+.content-element-bar .v-card__text {
+	margin-top: 0px;
+}
+.content-element-bar-element {
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementBar.vue]: fixed

File: src/components/ui-board/content-element/ContentElementDisplay.vue [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+import { VuetifyIcon } from "vuetify/types/services/icons";
+
+export default defineComponent({
+	name: "ContentElementDisplay",
+	props: {
+		hasGreyBackground: { type: Boolean, required: false },
+		icon: { type: String as PropType<VuetifyIcon>, required: false },
+	},
+	setup(props) {
+		const backgroundClass = computed(() => {
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementDisplay.vue]: fixed

File: src/components/ui-board/content-element/ContentElementDisplay.vue [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+</template>
+
+<script lang="ts">
+import { computed, defineComponent, PropType } from "vue";
+import { VuetifyIcon } from "vuetify/types/services/icons";
+
+export default defineComponent({
+	name: "ContentElementDisplay",
+	props: {
+		hasGreyBackground: { type: Boolean, required: false },
```
[the remark this answers is not in the document]
Comment (hoeppner-dataport) [reply src/components/ui-board/content-element/ContentElementDisplay.vue]: fixed
